Testing our apis (#11)
Ada-Developers-Academy/core-unit-2: testing-our-apis/environments-and-env-vars.md

# Environments and Env Vars

## Goals

Before we get to automated testing for our API, we'll need to learn about managing different environments.

> Our goal is to:
> 
> 1. Define what are development and testing environments
> 1. Define environment variables and manage them

Reviewer: revise the goals (plural, since more than one)
we don't go much into what "managing" means so I'd remove that from the second goal

```suggestion
Our goals are:

1. Define development and testing environments
1. Define environment variables
```

> 

Reviewer: if we still want to call out "manage" consider adding a dedicated third goal about managing
```suggestion
1. Start thinking about how to manage environment variables

```

## Introduction

Aleesha dances on a dance team. Aleesha ends up dancing in a lot of different places:

- She works on, practices, and learns how to dance at home
- She rehearses with her teammates at a dance studio, where they use a shared practice room
- She dances on stage at the big performances, where multiple teams and an audience share the space

Although Aleesha is _dancing_ in all of them, depending on her goal, she uses a different _environment_.

## Environments

In software, the word "environment" is used in many different ways. The meaning of "environment" changes based on the context.

### !callout-secondary

## Realistic Expectations

> No, really. "Environment" is a common concept among many different situations. When in doubt, clarify with questions!

Reviewer: saying it's a common concept sounds like it should have the same meaning everywhere
rephrase to emphasize the definition is overloaded
```suggestion
No, really. "Environment" can mean many different things in many different situations. When in doubt, clarify with questions!
```

### !end-callout

> In the context of local project development and testing, there are two common patterns:

Reviewer: this is showing a single pattern with two parts
```suggestion
In the context of local project development and testing, we will commonly use two environments according to the following pattern:
```

> 1. A _development environment_ dedicated for rapid development
> 1. A _testing environment_ dedicated for holding test data

Reviewer: 'to' sounds a little better to me and has agreement with the phrasing in the table
```suggestion
1. A _development environment_ dedicated to rapid development
1. A _testing environment_ dedicated to holding test data
```

> | Name                    | Purpose                                                                                        | Differences                          |

Reviewer: Differences sounds a little vague given the actual values in the table
```suggestion
| Name                    | Purpose                                                                                        | <div style="min-width:200px">Configuration Details</div>                          |
```

| ----------------------- | ---------------------------------------------------------------------------------------------- | ------------------------------------ |

> | Development environment | Dedicated to rapid development. Things will change often as the developer works on the project | Has a dedicated development database |

Reviewer: explain what's in the development database, and call out debugging/logging
```suggestion
| Development environment | Dedicated to rapid development. Things will change often as we work on the project | Has a dedicated development database where we can add sample data we will use during development. May enable a debugging mode with detailed logging and error reports. |
```

> | Testing environment     | Dedicated to setting up the data for running test suites.                                      | Has a dedicated testing database     |

Reviewer: what's special about the testing database?
```suggestion
| Testing environment     | Dedicated to running test suites.                                      | Has a dedicated testing database containing specific data used to exercise our test suite.     |
```

> As developers, we'll have access to both environments. Whenever we run our Flask API, we can designate it to use a specific environment.

Reviewer: we don't just _have_ access, we access them both a lot.
our discussions so far have made it sound like dev and test are completely distinct (and they can be), but since we'll be doing dev and test from our local machine, provide some explanation of why that is.
```suggestion
As developers, we access both environments.

As we develop our Flask API, we will work in our development environment. As we make changes to our code, we will test it frequently using our test environment. 

These two environments could exist on different machines. In practice, it is usually more convenient to have a way to run under either environment from our own developer machine. This allows us to test our code as we are working on it, minimizing the chance of breaking previous functionality (called a _regression_).

Whenever we run our Flask API, we can designate which environment to use.
```

> The environment that our Flask API uses will determine which database it accesses while it runs.

Reviewer: the environment can affect even more, so keep this at the level of general customization, but provide the database as an example of this
```suggestion
We will customize how we set certain configuration settings in our Flask API—such as the database it will access—based on the environment.
```

### Applying Environments: Dance Metaphor

When Aleesha dances, she picks an environment. Different environments are configured to use different resources.

Each environment has lighting. However, the lighting equipment doesn't, and shouldn't, be identical between the different environments:

- Her practice setup at home has some small LED lights
- The team's practice studio has adjustable ceiling lights
- The main performance stage has professional lighting equipment

> Keeping these environments separate can help her have better expectations for what's available.

Reviewer: having expectations about what's available sounds like we're bracing ourselves for disappointment. try focussing on the clarity and focus we gain by using different environments for different purposes.
```suggestion
Keeping these environments separate helps Aleesha and the other dancers focus on doing specific things in each environment.
```

> We can imagine Aleesha's practice room at home is like a _development environment_.

Reviewer: like I wrote above, we are basically using both dev and test together, so tie these concepts back to the metaphor.
```suggestion
We can imagine Aleesha's practice room at home is like her combined _development_ and _test_ environment.

As her _development_ environment, she can develop new dance routines in her own space. She can easily reconfigure the lighting and arrangement of things in that space to suit her work.

As her _test_ environment, she can practice and practice and make sure she has everything down before meeting with the rest of her dance team.
```

> 

Reviewer: Our metaphor has the studio and main stage parts as well. I'd like to explain where they fit into dev practice as well.
```suggestion

### !callout-info

## More Environments Are Possible

For now, we will focus on development and test environments. But there are other environments that are also useful.

<br />

We can have _staging_ environments, which are more like the practice studio used by Aleesha's whole team. It's a place where all the work they've done individually comes together and can be seen working as a whole. Since this is a shared space involving multiple people, it's harder to make changes here.

<br />

We will also have a _production_ environment. This is the main performance stage, where the rest of the world sees our hard work. Once the performance starts, we _really_ don't want to make changes!

### !end-callout

```

Author: YAYYYY staging!!! I wasn't brave enough to pull in staging but this is great!! Thanks!

Reviewer: It _could_ get confused with the test environment, but I think the practice studio you provided in the metaphor will do a great job of helping the reader keep these things separate in their mind, even though we won't really be using staging environments.

### Applying Environments at a High-Level

> As we introduce testing into our Flask API development, we will want to setup a dedicated testing environment. This will make our testing experience more consistent and reliable.

Reviewer: explain why we want this and how it makes the experience better
```suggestion
As we introduce testing into our Flask API development, we want it to be easy for our tests to run using our test environment settings.

Having dedicated test settings is important so that our tests can run under the actual conditions we want to test. Making this change automatic will help us avoid running tests using the wrong data.

This will make our testing experience more consistent and reliable.
```

## Environment Variables

> Environment variables are strings that hold data outside of our application code. These variables are set at the environment level, and affect how an app is run.

Reviewer: the jump from environments to variables feels disjoint, so included a small transition sentence.
as a reader _ don't know that I can identify with something being set  "at the environment level". I modified it to say that our environments will provide ways to set this (there's an added callout a bit later showing a small example)
give some overview of the kinds of affects they have on our code. (the details come in later bullets)

```suggestion
We will use environment variables to store information about values that change between environments.

Environment variables are strings that hold data outside of our application code. The environments we use provide ways to set these kinds of values.

Within our code, we can examine these variables, and either use them directly as data, or use them to make decisions about how our code should behave.
```

Some examples include:

> - Where the location of a database is
> - Where the location of other resources are (such as other services part of a larger software system)

Reviewer: don't need where (especially compared to the other bullets)
callback to connection string vocab
I kept reading part as a noun which was throwing my ability to parse the sentence
```suggestion
- The location of a database (a connection string)
- Locations of other resources (such as other services that are part of a larger software system)
```

- Various configurations about how to run the app, such as debugging mode
- API keys

> 

Reviewer: Add a callout showing a simple example of seeing and modifying environment variables
no suggestion because code
following comment includes the markdown

Reviewer: ### !callout-info

## There Are Many Environment Variables Set Already

From any terminal prompt, we can run the command `env`. This displays the environment variables currently set in our terminal. We can set our own environment variables from the terminal as well. Let's run the following:

```bash
$ export MY_NEW_VAR="my value"
```

<br />

If we then run `env` again, we will see our environment variable `MY_NEW_VAR` appear among the output.

### !end-callout

### Managing Environment Variables

> Environment variables may hold a lot of values related to configuration.
> 
> Additionally, environment variables often hold _sensitive and secret data_.
> 
> In both of these cases, we will need to manage them. Managing environment variables lets us:
> 
> - Keep them organized together, rather than as constant variables throughout the project code
> - Control their visibility
> 
> It's often the case that we _must_ keep environment variables secret to the public. Because our codebases are often online, it may be necessary to **not commit environment variables to source control**.
> 
> Different tools, packages, and practices will give us ways to manage and hide our environment variables.

Reviewer: wound reorganizing this a bit
maintain the emphasis on avoiding committing
built a bit of a story about how we got from "values outside our code" to something we might accidentally commit.
included a callout emphasizing that we will be looking at just a small part of what goes into configuring environments.

```suggestion
Because environment variables are held outside of our code, they do not get checked into source control. This makes them very useful for storing _sensitive and secret data_.

- We might need a username and password to connect to a database.
- An API key is usually unique to our account. If someone else gets it, they can misuse a service acting under our identity.

As our applications grow, we may need to store a larger number of variables, to the point that using the `export` command becomes a problem. It would be nice if we could group our environment variables into a special file: a file that holds only our environment variables, and no code.

But once we have placed them in a file, we might accidentally commit that file. Because of the danger of misuse, we want to **avoid committing environment variables to source control**, where nefarious individuals could steal our credentials!

It seems like we're right back where we started!

In the next lesson, will explore how to manage our environment variables so that they remain organized, out of our code, and secure!

### !callout-info

## Many Ways to Customize Environments

Flask doesn't require a particular approach to setting environment configurations. We will keep our discussion of configuration focused on switching between databases and storing application secrets. Customizing configuration is a much larger topic, with many clever approaches to explore. Follow your curiosity!

### !end-callout
```
